Assets/Scripts/Combat/Abilities/MagicShield.cs

```diff
@@ -0,0 +1,269 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+public class MagicShield : Ability
+{
+    private enum Button { Up, Down, Left, Right };
+    private enum Phase { SequencePhase, InputSequence, Activate, Inactive };
+    private Timer SequenceTimer;
+    private Queue<Button> Sequence = new Queue<Button>();
+    private Queue<Button> SequenceOrder = new Queue<Button>();
+    private Queue<GameObject> Directions = new Queue<GameObject>();
+
+    [SerializeField]
+    private GameObject Shield;
+    [SerializeField]
+    private GameObject[] Arrows; // 0: Up, 1: Down, 2: Left, 3: Right 
+
+    private readonly float SequenceDuration = 2f;
+    private readonly float InputDuration = 5f;
+    private readonly float PositionOffset = 4f;
+    private Vector3 Position = Vector3.zero;
+    private Phase CurrentPhase = Phase.Inactive;
+
+    private int CorrectInputs = 0;
+    private readonly int MagicShieldFactor = 20;
+    private int MagicShieldHealth;
+
+    public new void Start()
+    {
+        base.Start();
+        SequenceTimer = GetComponent<Timer>();
+        Shield.SetActive(false);
+
+        foreach(GameObject arrow in Arrows)
+            arrow.SetActive(false);
+
+        TargetSchema = new TargetSchema(
+            0,
+            CombatantType.Ally,
+            SelectorType.All);
+    }
+
+    new public void StartAbility()
+    {
+        CorrectInputs = 0;
+        base.StartAbility();
+    }
+
+    private void Update()
+    {
+        if(CurrentPhase == Phase.SequencePhase)
+            SequencePhaseUpdate();
+
+        else if(CurrentPhase == Phase.InputSequence)
+            InputSequenceUpdate();
+    }
+
+    protected override void ContinueAbilityAfterTargeting()
+    {
+        StartMagicShieldSequence();
+    }
+
+    protected override void EndAbility()
+    {
+        Debug.Log($"Magic Shield Total Health: {MagicShieldHealth}");
+
+        for (int i = 0; i < TargetedCombatants.Length; i++)
+            TargetedCombatants[i].GetComponent<Combatant>().ApplyShield(MagicShieldHealth);
+
+        CombatSystem.EndTurn(this.GetComponentInParent<Combatant>().gameObject);
+
+    }
+
+    private void StartMagicShieldSequence()
+    {
+        Debug.Log("Starting Magic Shield Ability");
+
+        Shield.transform.position = GameObject.FindGameObjectWithTag("Player").transform.position;
+        Shield.transform.position -= new Vector3(0, 5, 0);
+        Shield.transform.Translate(new Vector3(0, 10, 0));
+
+        Array buttons = Enum.GetValues(typeof(Button));
+
+        for (int i = 0; i < buttons.Length; i++)
+        {
+            Button random_button = (Button)buttons.GetValue(Random.Range(0, buttons.Length));
+            Sequence.Enqueue(random_button);
+            SequenceOrder.Enqueue(random_button);
+        }
```

Comment:
I think it may have been mentioned, but just in case, you can instead just push the randomly selected button at the end of the sequence, and then you won't need two queues anymore. I would maybe rename `Sequence` to `InputSequence` or something alike.

Reply:
I just refactored it now. What I did now is removed the two queues and just have a single list. This single list will act like a pseudo Queue. If I remove at the very first index, it will shift all the elements to the front and when I add an element, it will add it at the end of the list. This makes it a lot easier too since I don't have to dequeue something when I want an element. I'll just "dequeue" something if I want the next element in the list and if I need it to check the sequence again, then I'll just add it back to the end of the list. Once I go through the whole list, everything will be back to how it was.
It's sorta what you said here, but I'm using a list instead.